Сравнение музыкальных пристрастий Москвы и Петербурга


На данных вы сравните поведение пользователей двух столиц.

**Цель исследования** — проверьте три гипотезы:
1. Активность пользователей зависит от дня недели. Причём в Москве и Петербурге это проявляется по-разному.
2. В понедельник утром в Москве преобладают одни жанры, а в Петербурге — другие. Так же и вечером пятницы преобладают разные жанры — в зависимости от города. 
3. Москва и Петербург предпочитают разные жанры музыки. В Москве чаще слушают поп-музыку, в Петербурге — русский рэп.

**Ход исследования**

Данные о поведении пользователей вы получите из файла `music_project.csv`. О качестве данных ничего не известно. Поэтому перед проверкой гипотез понадобится обзор данных. 

Вы проверите данные на ошибки и оцените их влияние на исследование. Затем, на этапе предобработки вы поищете возможность исправить самые критичные ошибки данных.
 
Таким образом, исследование пройдёт в три этапа:
 1. Обзор данных.
 2. Предобработка данных.
 3. Проверка гипотез.
 
 
**Согласно документации к данным:**

userID — идентификатор пользователя

Track — название трека

artist — имя исполнителя

genre — название жанра

City — город пользователя

time — время начала прослушивания

Day — день недели.

## Обзор данных

**1. Импортируйте эту библиотеку `pandas`**

In [1]:
# импорт библиотеки pandas
import pandas as pd

**2. Прочитайте файл music_project.csv и сохраните его в переменной df**

**- Скачайте файл music_project.csv из LMS и скопируйте путь к файлу на локальном компьютере**


In [2]:
# чтение файла с данными и сохранение в df
df = pd.read_csv(r"/Users/eugeneinsidious/Downloads/music_project.csv") #аргумент - путь к файлу
df

userID                              Track            artist  \
0      FFB692EC                  Kamigata To Boots  The Mass Missile   
1      55204538        Delayed Because of Accident  Andreas Rönnberg   
2        20EC38                  Funiculì funiculà       Mario Lanza   
3      A3DD03C9              Dragons in the Sunset        Fire + Ice   
4      E2DC1FAE                        Soul People        Space Echo   
...         ...                                ...               ...   
65074  729CBB09                            My Name            McLean   
65075  D08D4A55  Maybe One Day (feat. Black Spade)       Blu & Exile   
65076  C5E3A0D5                          Jalopiina               NaN   
65077  321D0506                      Freight Train     Chas McDevitt   
65078  3A64EF84          Tell Me Sweet Little Lies      Monica Lopez   

            genre            City        time        Day  
0            rock  Saint-Petersburg  20:28:33  Wednesday  
1            rock            Moscow  14:07:09     Friday  
2             pop  Saint-Petersburg  20:58:07  Wednesday  
3            folk  Saint-Petersburg  08:37:09     Monday  
4           dance            Moscow  08:34:34     Monday  
...           ...               ...       ...        ...  
65074         rnb            Moscow  13:32:28  Wednesday  
65075         hip  Saint-Petersburg  10:00:00     Monday  
65076  industrial            Moscow  20:09:26     Friday  
65077        rock            Moscow  21:43:59     Friday  
65078     country            Moscow  21:59:46     Friday  

[65079 rows x 7 columns]

**3. Выведите на экран первые десять строк таблицы:**

In [3]:
# получение первых 10 строк таблицы df
df.head(10)

userID                        Track            artist   genre  \
0  FFB692EC            Kamigata To Boots  The Mass Missile    rock   
1  55204538  Delayed Because of Accident  Andreas Rönnberg    rock   
2    20EC38            Funiculì funiculà       Mario Lanza     pop   
3  A3DD03C9        Dragons in the Sunset        Fire + Ice    folk   
4  E2DC1FAE                  Soul People        Space Echo   dance   
5  842029A1                    Преданная         IMPERVTOR  rusrap   
6  4CB90AA5                         True      Roman Messer   dance   
7  F03E1C1F             Feeling This Way   Polina Griffith   dance   
8  8FA1D3BE     И вновь продолжается бой               NaN  ruspop   
9  E772D5C0                    Pessimist               NaN   dance   

             City        time        Day  
0  Saint-Petersburg  20:28:33  Wednesday  
1            Moscow  14:07:09     Friday  
2  Saint-Petersburg  20:58:07  Wednesday  
3  Saint-Petersburg  08:37:09     Monday  
4            Moscow  08:34:34     Monday  
5  Saint-Petersburg  13:09:41     Friday  
6            Moscow  13:00:07  Wednesday  
7            Moscow  20:47:49  Wednesday  
8            Moscow  09:17:40     Friday  
9  Saint-Petersburg  21:20:49  Wednesday

**4. Одной командой получить общую информацию о таблице:**

In [4]:
# получение общей информации о данных в таблице df
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 65079 entries, 0 to 65078
Data columns (total 7 columns):
 #   Column    Non-Null Count  Dtype 
---  ------    --------------  ----- 
 0     userID  65079 non-null  object
 1   Track     63848 non-null  object
 2   artist    57876 non-null  object
 3   genre     63881 non-null  object
 4     City    65079 non-null  object
 5   time      65079 non-null  object
 6   Day       65079 non-null  object
dtypes: object(7)
memory usage: 3.5+ MB


**Ваши выводы по обзору данных**













Имеем таблицу размером 65079 строк, столбцов 7
Тип данных в колонках: object
В столбцах(4): userID, City, time, Day - одинаковое количество элементов (65079)
А в столбцах(3): Track, artist, genre - разное количество элементов
В каждом из этих 3х столбцов встречаются "нулевые", пропущенные значения. Их нужно будет обработать
Есть подозрения, что допущены лишние пробелы/неверный стиль в столбцах userID, City, в остальных то нижний регистр, то верхний регистр, наверное нужен единый стиль оформления

## Предобработка данных
Исправьте стиль в заголовках столбцов, исключите пропуски. Затем проверьте данные на дубликаты.

**5. Выведите на экран названия столбцов:**

In [5]:
# перечень названий столбцов таблицы df
print(df.columns)

Index(['  userID', 'Track', 'artist', 'genre', '  City  ', 'time', 'Day'], dtype='object')


**6. переименование столбцов**

In [6]:
# переименование столбцов
             
df = df.rename(columns={'  userID':'user_id', 'Track':'track', 'artist':'artist', 'genre':'genre', '  City  ':'city', 'time':'start_time', 'Day':'weekday'})
df.columns        

Index(['user_id', 'track', 'artist', 'genre', 'city', 'start_time', 'weekday'], dtype='object')

**7. Проверьте результат. Для этого ещё раз выведите на экран названия столбцов:**

In [7]:
# проверка результатов - перечень названий столбцов
print(df.columns)

Index(['user_id', 'track', 'artist', 'genre', 'city', 'start_time', 'weekday'], dtype='object')


### Пропуски значений
Сначала посчитайте, сколько в таблице пропущенных значений. Для этого достаточно двух методов `pandas`:

**8. подсчёт пропусков**

In [8]:
# подсчёт пропусков
print(df.isna().sum())

user_id          0
track         1231
artist        7203
genre         1198
city             0
start_time       0
weekday          0
dtype: int64


Не все пропущенные значения влияют на исследование. Так в `track` и `artist` пропуски не важны для нашей работы. Достаточно заменить их явными обозначениями.

Но пропуски в `genre` могут помешать сравнению музыкальных вкусов в Москве и Санкт-Петербурге. На практике было бы правильно установить причину пропусков и восстановить данные. Такой возможности нет. 

Придётся:
* заполнить и эти пропуски явными обозначениями,
* оценить, насколько они повредят расчётам. 

**9. Замените пропущенные значения в столбцах `track`, `artist` и `genre` на строку `'unknown'`. Для этого создайте список `columns_to_replace`, переберите его элементы циклом `for` и для каждого столбца выполните замену пропущенных значений:**

In [9]:
# перебор названий столбцов в цикле и замена пропущенных значений на 'unknown'
columns_to_replace = ['track', 'artist', 'genre']

for i in columns_to_replace:
    df[i] = df[i].fillna('unknown')


**10. Убедитесь, что в таблице не осталось пропусков. Для этого ещё раз посчитайте пропущенные значения.**

In [10]:
# подсчёт пропусков
print(df.isna().sum())

user_id       0
track         0
artist        0
genre         0
city          0
start_time    0
weekday       0
dtype: int64


### Дубликаты

**11. Посчитайте явные дубликаты в таблице одной командой:**

In [11]:
# подсчёт явных дубликатов
print(df.duplicated().sum())

3826


**12. Вызовите специальный метод `pandas`, чтобы удалить явные дубликаты:**

In [39]:
# удаление явных дубликатов (с удалением старых индексов и формированием новых)
df = df.drop_duplicates().reset_index(drop=True)

**13. Ещё раз посчитайте явные дубликаты в таблице — убедитесь, что полностью от них избавились:**

In [40]:
# проверка на отсутствие дубликатов
print(df.duplicated().sum())

0


Теперь избавьтесь от неявных дубликатов в колонке `genre`. Например, название одного и того же жанра может быть записано немного по-разному. Такие ошибки тоже повлияют на результат исследования.

**14. Выведите на экран список уникальных названий жанров, отсортированный в алфавитном порядке. Для этого:**

**- извлеките нужный столбец датафрейма,**

**- примените к нему метод сортировки,**

**- для отсортированного столбца вызовите метод, который вернёт уникальные значения из столбца.**

In [14]:
# Просмотр уникальных названий жанров

df_sorted = df.sort_values('genre')

print(df_sorted['genre'].unique())

['acid' 'acoustic' 'action' 'adult' 'africa' 'afrikaans' 'alternative'
 'alternativepunk' 'ambient' 'americana' 'animated' 'anime' 'arabesk'
 'arabic' 'arena' 'argentinetango' 'art' 'audiobook' 'author' 'avantgarde'
 'axé' 'baile' 'balkan' 'beats' 'bigroom' 'black' 'bluegrass' 'blues'
 'bollywood' 'bossa' 'brazilian' 'breakbeat' 'breaks' 'broadway'
 'cantautori' 'cantopop' 'canzone' 'caribbean' 'caucasian' 'celtic'
 'chamber' 'chanson' 'children' 'chill' 'chinese' 'choral' 'christian'
 'christmas' 'classical' 'classicmetal' 'club' 'colombian' 'comedy'
 'conjazz' 'contemporary' 'country' 'cuban' 'dance' 'dancehall' 'dancepop'
 'dark' 'death' 'deep' 'deutschrock' 'deutschspr' 'dirty' 'disco' 'dnb'
 'documentary' 'downbeat' 'downtempo' 'drum' 'dub' 'dubstep' 'eastern'
 'easy' 'electronic' 'electropop' 'emo' 'entehno' 'epicmetal' 'estrada'
 'ethnic' 'eurofolk' 'european' 'experimental' 'extrememetal' 'fado'
 'fairytail' 'film' 'fitness' 'flamenco' 'folk' 'folklore' 'folkmetal'
 'folkrock' 

**15. Просмотрите список и найдите неявные дубликаты названия `hiphop`. Это могут быть названия с ошибками или альтернативные названия того же жанра.**

Вы увидите следующие неявные дубликаты:
* *hip*,
* *hop*,
* *hip-hop*.

Чтобы очистить от них таблицу, напишите функцию `replace_wrong_genres()` с двумя параметрами: 
* `wrong_genres` — список дубликатов,
* `correct_genre` — строка с правильным значением.

Функция должна исправить колонку `genre` в таблице `df`: заменить каждое значение из списка `wrong_genres` на значение из `correct_genre`.

In [15]:
# Функция для замены неявных дубликатов

def replace_wrong_genres(wrong_genres, correct_genre):
    for wrong_genre in wrong_genres: 
        df['genre'] = df['genre'].replace(wrong_genres, correct_genre)

wrong_genres = ['hip', 'hip-hop', 'hop']
correct_genre = 'hiphop'  

**16. Вызовите `replace_wrong_genres()` и передайте ей такие аргументы, чтобы она устранила неявные дубликаты: вместо `hip`, `hop` и `hip-hop` в таблице должно быть значение `hiphop`:**

In [16]:
# Устранение неявных дубликатов

replace_wrong_genres(wrong_genres, correct_genre) 


**17. Проверка на неявные дубликаты**

In [17]:
# Проверка на неявные дубликаты
df_sorted = df.sort_values('genre')
print(df_sorted['genre'].unique())

['acid' 'acoustic' 'action' 'adult' 'africa' 'afrikaans' 'alternative'
 'alternativepunk' 'ambient' 'americana' 'animated' 'anime' 'arabesk'
 'arabic' 'arena' 'argentinetango' 'art' 'audiobook' 'author' 'avantgarde'
 'axé' 'baile' 'balkan' 'beats' 'bigroom' 'black' 'bluegrass' 'blues'
 'bollywood' 'bossa' 'brazilian' 'breakbeat' 'breaks' 'broadway'
 'cantautori' 'cantopop' 'canzone' 'caribbean' 'caucasian' 'celtic'
 'chamber' 'chanson' 'children' 'chill' 'chinese' 'choral' 'christian'
 'christmas' 'classical' 'classicmetal' 'club' 'colombian' 'comedy'
 'conjazz' 'contemporary' 'country' 'cuban' 'dance' 'dancehall' 'dancepop'
 'dark' 'death' 'deep' 'deutschrock' 'deutschspr' 'dirty' 'disco' 'dnb'
 'documentary' 'downbeat' 'downtempo' 'drum' 'dub' 'dubstep' 'eastern'
 'easy' 'electronic' 'electropop' 'emo' 'entehno' 'epicmetal' 'estrada'
 'ethnic' 'eurofolk' 'european' 'experimental' 'extrememetal' 'fado'
 'fairytail' 'film' 'fitness' 'flamenco' 'folk' 'folklore' 'folkmetal'
 'folkrock' 

**Ваши выводы по предобработке данных**







На этапе предобработки в данных обнаружились пропуски, дубликаты, а также неверное форматирование стиля в заголовках столбцов. Получилось устранить пропуски и дубликаты, изменить стиль заголовков, теперь можно начать исследование 

## Проверка гипотезы 1

**18. Оцените активность пользователей в каждом городе. Сгруппируйте данные по городу и посчитайте прослушивания в каждой группе.**

In [21]:
# Подсчёт прослушиваний в каждом городе
print(df.groupby('city')['genre'].count())

city
Moscow              42741
Saint-Petersburg    18512
Name: genre, dtype: int64


**19. Теперь сгруппируйте данные по дню недели и подсчитайте прослушивания в понедельник, среду и пятницу.**

In [22]:
# Подсчёт прослушиваний в каждый из трёх дней
print(df.groupby('weekday')['genre'].count())

weekday
Friday       21840
Monday       21354
Wednesday    18059
Name: genre, dtype: int64


**20 Создайте функцию number_tracks(), которая посчитает прослушивания для заданного дня и города. Ей понадобятся два параметра:**

- день недели

- название города.

В функции сохраните в переменную строки исходной таблицы, у которых значение:

в колонке day равно параметру day,
в колонке city равно параметру city.
Для этого примените последовательную фильтрацию с логической индексацией.

Затем посчитайте значения в столбце user_id получившейся таблицы. Результат сохраните в новую переменную. Верните эту переменную из функции.

In [25]:
def number_tracks(day, city):
    track_list = df.loc[df['weekday'] == day] 
    track_list = track_list[track_list['city'] == city]
    track_list_count = track_list['user_id'].count()
    return track_list_count


# <создание функции number_tracks()>
# Объявляется функция с двумя параметрами: day, city.
# В переменной track_list сохраняются те строки таблицы df, для которых 
# значение в столбце 'day' равно параметру day и одновременно значение
# в столбце 'city' равно параметру city (используйте последовательную фильтрацию
# с помощью логической индексации).
# В переменной track_list_count сохраняется число значений столбца 'user_id',
# рассчитанное методом count() для таблицы track_list.
# Функция возвращает число - значение track_list_count.

# Функция для подсчёта прослушиваний для конкретного города и дня.
# С помощью последовательной фильтрации с логической индексацией она 
# сначала получит из исходной таблицы строки с нужным днём,
# затем из результата отфильтрует строки с нужным городом,
# методом count() посчитает количество значений в колонке user_id. 
# Это количество функция вернёт в качестве результата

**21. Вызовите number_tracks() шесть раз, меняя значение параметров — так, чтобы получить данные для каждого города в каждый из трёх дней.**

In [26]:
# количество прослушиваний в Москве по понедельникам

number_tracks('Monday', 'Moscow')

15740

In [27]:
# количество прослушиваний в Санкт-Петербурге по понедельникам

number_tracks('Monday', 'Saint-Petersburg')

5614

In [28]:
# количество прослушиваний в Москве по средам

number_tracks('Wednesday', 'Moscow')

11056

In [29]:
# количество прослушиваний в Санкт-Петербурге по средам

number_tracks('Wednesday', 'Saint-Petersburg')

7003

In [30]:
# количество прослушиваний в Москве по пятницам

number_tracks('Friday', 'Moscow')

15945

In [31]:
# количество прослушиваний в Санкт-Петербурге по пятницам

number_tracks('Friday', 'Saint-Petersburg')

5895

**22. Создайте c помощью конструктора pd.DataFrame таблицу, где**

- названия колонок — ['city', 'monday', 'wednesday', 'friday']

- данные — результаты, которые вы получили с помощью number_tracks.

In [33]:
data = [
    ['Moscow', 15740, 11056, 15945],
    ['Saint-Petersburg', 5614, 7003, 5895]
]

columns = ['city', 'monday', 'wednesday', 'friday']
table = pd.DataFrame(data=data, columns=columns) 
table 

city  monday  wednesday  friday
0            Moscow   15740      11056   15945
1  Saint-Petersburg    5614       7003    5895

**Выводы по первой гипотезе**






Активность пользователей зависит от дня недели. 
В Москве пик активности приходится на пятницу, в понедельник чуть меньше пятницы, а среду активность снижается.
В Петербурге наоборот, пиковая активнось приходится на среду, в пятницу и понедельник активность снижается, понедельник чуть-чуть уступает пятнице.

## Проверка гипотезы 2

**23. Сохраните таблицы с данными в две переменные:**
* по Москве — в `moscow_general`;
* по Санкт-Петербургу — в `spb_general`.

In [41]:
# получение таблицы moscow_general из тех строк таблицы df, 
# для которых значение в столбце 'city' равно 'Moscow'

moscow_general = df[df['city'] == 'Moscow']

moscow_general

user_id                          track                   artist  \
1      55204538    Delayed Because of Accident         Andreas Rönnberg   
4      E2DC1FAE                    Soul People               Space Echo   
6      4CB90AA5                           True             Roman Messer   
7      F03E1C1F               Feeling This Way          Polina Griffith   
8      8FA1D3BE       И вновь продолжается бой                  unknown   
...         ...                            ...                      ...   
61247  83A474E7  I Worship Only What You Bleed  The Black Dahlia Murder   
61248  729CBB09                        My Name                   McLean   
61250  C5E3A0D5                      Jalopiina                  unknown   
61251  321D0506                  Freight Train            Chas McDevitt   
61252  3A64EF84      Tell Me Sweet Little Lies             Monica Lopez   

              genre    city start_time    weekday  
1              rock  Moscow   14:07:09     Friday  
4             dance  Moscow   08:34:34     Monday  
6             dance  Moscow   13:00:07  Wednesday  
7             dance  Moscow   20:47:49  Wednesday  
8            ruspop  Moscow   09:17:40     Friday  
...             ...     ...        ...        ...  
61247  extrememetal  Moscow   21:07:12     Monday  
61248           rnb  Moscow   13:32:28  Wednesday  
61250    industrial  Moscow   20:09:26     Friday  
61251          rock  Moscow   21:43:59     Friday  
61252       country  Moscow   21:59:46     Friday  

[42741 rows x 7 columns]

In [36]:
# получение таблицы spb_general из тех строк таблицы df,
# для которых значение в столбце 'city' равно 'Saint-Petersburg'

spb_general = df[df['city'] == 'Saint-Petersburg']

spb_general

user_id                              track              artist  \
0      FFB692EC                  Kamigata To Boots    The Mass Missile   
2        20EC38                  Funiculì funiculà         Mario Lanza   
3      A3DD03C9              Dragons in the Sunset          Fire + Ice   
5      842029A1                          Преданная           IMPERVTOR   
9      E772D5C0                          Pessimist             unknown   
...         ...                                ...                 ...   
61239  D94F810B        Theme from the Walking Dead  Proyecto Halloween   
61240  BC8EC5CF       Red Lips: Gta (Rover Rework)               Rover   
61241  29E04611                       Bre Petrunko       Perunika Trio   
61242  1B91C621             (Hello) Cloud Mountain     sleepmakeswaves   
61249  D08D4A55  Maybe One Day (feat. Black Spade)         Blu & Exile   

            genre              city start_time    weekday  
0            rock  Saint-Petersburg   20:28:33  Wednesday  
2             pop  Saint-Petersburg   20:58:07  Wednesday  
3            folk  Saint-Petersburg   08:37:09     Monday  
5          rusrap  Saint-Petersburg   13:09:41     Friday  
9           dance  Saint-Petersburg   21:20:49  Wednesday  
...           ...               ...        ...        ...  
61239        film  Saint-Petersburg   21:14:40     Monday  
61240  electronic  Saint-Petersburg   21:06:50     Monday  
61241       world  Saint-Petersburg   13:56:00     Monday  
61242    postrock  Saint-Petersburg   09:22:13     Monday  
61249      hiphop  Saint-Petersburg   10:00:00     Monday  

[18512 rows x 7 columns]

**24. Создайте функцию genre_weekday() с четырьмя параметрами:**

* таблица (датафрейм) с данными,
* день недели,
* начальная временная метка в формате 'hh:mm',
* последняя временная метка в формате 'hh:mm'.

Функция должна вернуть информацию о топ-10 жанров тех треков, которые прослушивали в указанный день, в промежутке между двумя отметками времени.

In [44]:
def genre_weekday(table, day, time1, time2):
    genre_df = table[table['weekday'] == day]
    genre_df = genre_df[genre_df['start_time'] > time1]
    genre_df = genre_df[genre_df['start_time'] < time2]
    genre_df_count = genre_df.groupby('genre')['genre'].count()
    genre_df_sorted = genre_df_count.sort_values(ascending=False)
    return genre_df_sorted.head(10)


# Объявление функции genre_weekday() с параметрами table, day, time1, time2,
# которая возвращает информацию о самых популярных жанрах в указанный день в
# заданное время:
# 1) в переменную genre_df сохраняются те строки переданного датафрейма table, для
#    которых одновременно:
#    - значение в столбце day равно значению аргумента day
#    - значение в столбце time больше значения аргумента time1
#    - значение в столбце time меньше значения аргумента time2
#    Используйте последовательную фильтрацию с помощью логической индексации.
# 2) сгруппировать датафрейм genre_df по столбцу genre, взять один из его
#    столбцов и посчитать методом count() количество записей для каждого из
#    присутствующих жанров, получившийся Series записать в переменную
#    genre_df_count
# 3) отсортировать genre_df_count по убыванию встречаемости и сохранить
#    в переменную genre_df_sorted
# 4) вернуть Series из 10 первых значений genre_df_sorted, это будут топ-10
#    популярных жанров (в указанный день, в заданное время)

**25. Cравните результаты функции genre_weekday() для Москвы и Санкт-Петербурга в понедельник утром (с 7:00 до 11:00) и в пятницу вечером (с 17:00 до 23:00):**

In [45]:
# вызов функции для утра понедельника в Москве (вместо df — таблица из задания 23)
# объекты, хранящие время, являются строками и сравниваются как строки
# пример вызова: genre_weekday(таблица из задания 23, 'Monday', '07:00', '11:00')

genre_weekday(moscow_general, 'Monday', '07:00', '11:00')

genre
pop            781
dance          549
electronic     480
rock           474
hiphop         286
ruspop         186
world          181
rusrap         175
alternative    164
unknown        161
Name: genre, dtype: int64

In [46]:
# вызов функции для утра понедельника в Петербурге

genre_weekday(spb_general, 'Monday', '07:00', '11:00')

genre
pop            218
dance          182
rock           162
electronic     147
hiphop          80
ruspop          64
alternative     58
rusrap          55
jazz            44
classical       40
Name: genre, dtype: int64

In [47]:
# вызов функции для вечера пятницы в Москве

genre_weekday(moscow_general, 'Friday', '17:00', '23:00')

genre
pop            713
rock           517
dance          495
electronic     482
hiphop         273
world          208
ruspop         170
alternative    163
classical      163
rusrap         142
Name: genre, dtype: int64

In [48]:
# вызов функции для вечера пятницы в Петербурге

genre_weekday(spb_general, 'Friday', '17:00', '23:00')

genre
pop            256
electronic     216
rock           216
dance          210
hiphop          97
alternative     63
jazz            61
classical       60
rusrap          59
world           54
Name: genre, dtype: int64

**Выводы по второй гипотезе**

В понедельник утром в Москве и Петербурге преобладают одинаковые жанры, и там и там в топ-2 pop и dance. Но с 3-его места в топе идут различия: в топ 10 в Петербурге входит джаз и классическая музыка, в Москве эти жанры не входят в топ, вместо них есть жанр world и на 10-м месте неизвестный жанр.

В пятницу вечером в Москве и Петербурге жанр pop всё так же в топ-1, со второй строчки идут различия: в топ-10 в Петербурге есть джаз, но нет ruspop, а в Москве в топ-10 входит ruspop, но отсуствует джаз, все остальные жанры есть в двух городах, но занимают разные места в топе.


## Проверка гипотезы 3

**26. Сгруппируйте таблицу moscow по жанру и посчитайте прослушивания треков каждого жанра методом count(). Затем отсортируйте результат в порядке убывания и сохраните его в таблице moscow_genres.**

In [52]:
# одной строкой: группировка таблицы moscow по столбцу 'genre', 
# подсчёт числа значений 'genre' в этой группировке методом count(), 
# сортировка получившегося Series в порядке убывания и сохранение в moscow_genres

moscow_genres = moscow_general.groupby('genre')['genre'].count().sort_values(ascending=False)


**27. Выведите на экран первые десять строк moscow_genres:**

In [53]:
# просмотр первых 10 строк moscow_genres

moscow_genres.head(10)

genre
pop            5892
dance          4435
rock           3965
electronic     3786
hiphop         2096
classical      1616
world          1432
alternative    1379
ruspop         1372
rusrap         1161
Name: genre, dtype: int64

**28. Теперь повторите то же и для Петербурга.**

In [54]:
# одной строкой: группировка таблицы spb по столбцу 'genre', 
# подсчёт числа значений 'genre' в этой группировке методом count(), 
# сортировка получившегося Series в порядке убывания и сохранение в spb_genres

spb_genres = spb_general.groupby('genre')['genre'].count().sort_values(ascending=False)

**29. Выведите на экран первые десять строк spb_genres:**

In [55]:
# просмотр первых 10 строк spb_genres
spb_genres.head(10)

genre
pop            2431
dance          1932
rock           1879
electronic     1736
hiphop          960
alternative     649
classical       646
rusrap          564
ruspop          538
world           515
Name: genre, dtype: int64

**Выводы по третьей гипотезе**






В Москве и Петербурге в топ-5 входят одинаковые жанры музыки, позиции совпадают и там и там. На первом месте pop. С 6 места позиции отличаются, но жанры всё те же в обоих городах. В итоге Москва и Петербург предпочитают одинаковые жанры музыки. 

**Итоговый вывод по исследованию**







У нас было три гипотезы:
1. Активность пользователей зависит от дня недели. Причём в Москве и Петербурге это проявляется по-разному.
2. В понедельник утром в Москве преобладают одни жанры, а в Петербурге — другие. Так же и вечером пятницы преобладают разные жанры — в зависимости от города.
3. Москва и Петербург предпочитают разные жанры музыки. В Москве чаще слушают поп-музыку, в Петербурге — русский рэп.

В итоге подтвердилась только первая гипотеза (в Москве пик активности приходится на пятницу, в Петербурге пиковая активнось приходится на среду), а вторая и третья гипотезы оказались неверными. В понедельник утром в Москве и Петербурге преобладают одинаковые жанры, в пятницу вечером в Москве и Петербурге жанр pop всё так же в топ-1. Москва и Петербург предпочитают одинаковые жанры музыки. В Москве и Петербурге чаще слушают pop, dance, rock electronic и hiphop музыку. 
   

